### ISSUE TITLE
node process gets killed and all instances fails

### ISSUE THREAD

**Issue by alibehroozi** [Azure/azure-functions-nodejs-worker#508]
After some time that our functions are running and giving responses to the requests, the process of all functions gets killed.

This behavior is impacting our performance as requests should wait again for a cold start for more than a minute so nothing works.
this is the error we get in application insights:
```
Microsoft.Azure.WebJobs.Script.Workers.WorkerProcessExitException: node exited with code -1073740791
```

with this call stack:
```
Microsoft.Azure.WebJobs.Host.FunctionInvocationException:
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<ExecuteWithLoggingAsync>d__21.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.31.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:343)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<TryExecuteAsync>d__16.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.31.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:105)
Inner exception Microsoft.Azure.WebJobs.Script.Workers.WorkerProcessExitException handled at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw:
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Script.Description.WorkerFunctionInvoker+<InvokeCore>d__9.MoveNext (Microsoft.Azure.WebJobs.Script, Version=3.3.0.0, Culture=neutral, PublicKeyToken=null: D:\a\1\s\src\WebJobs.Script\Description\Workers\WorkerFunctionInvoker.cs:96)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Script.Description.FunctionInvokerBase+<Invoke>d__24.MoveNext (Microsoft.Azure.WebJobs.Script, Version=3.3.0.0, Culture=neutral, PublicKeyToken=null: D:\a\1\s\src\WebJobs.Script\Description\FunctionInvokerBase.cs:82)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Script.Description.FunctionGenerator+<Coerce>d__3`1.MoveNext (Microsoft.Azure.WebJobs.Script, Version=3.3.0.0, Culture=neutral, PublicKeyToken=null: D:\a\1\s\src\WebJobs.Script\Description\FunctionGenerator.cs:225)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2+<InvokeAsync>d__10.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.31.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionInvoker.cs:52)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<InvokeWithTimeoutAsync>d__28.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.31.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:572)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<ExecuteWithWatchersAsync>d__27.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.31.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:518)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<ExecuteWithLoggingAsync>d__21.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.31.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:296)
```

So do you have any idea why this is happening?

**Comment by alibehroozi** [comment]
any idea?

**Comment by ejizba** [comment]
Hi @alibehroozi please take a look at these instructions so that we can get information about your specific function app:
https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately

**Comment by alibehroozi** [comment]
sorry for the late reply.
Here is what I got from our Application Insights:

time: `2022-01-08T03:56:00.144Z`
region: `Central US`
InvocationId: `3f629c56-f21c-4db6-a31d-19df6c749710`
HostInstanceId: `e5df6581-ed57-4eae-aabe-9719a08ba4c2`

Let me know if any other information is needed.

**Comment by phillipharding** [comment]
I also had this issue a couple of days ago

A queue triggered function running correctly and had been since start up at the beginning of the day, then for no apparent reason the NodeJs process just terminates with an error in the log
> Microsoft.Azure.WebJobs.Script.Workers.WorkerProcessExitException: node exited with code -1073740791

After cold restart, replayed the original queue message, triggering the function which ran correctly and as expected.

|   |   |
| ------- | ------ |
| Function Worker | NodeJs |
| Runtime Version | ~3 |
| Node | v12 |
| Plan | Consumption |
| Region | UK South |

**Comment by EwaldBorger** [comment]
I'm not sure if this is the exact same issue (if not I'll open a new issue), but it looks like it.
We just see processes die and get no output at all about what caused it.

```
LanguageWorkerConsoleLog[error] Worker 45beffcf-867a-4f37-9e4d-01a8648ea725 encountered event stream error:  Error: 14 UNAVAILABLE: Connection dropped
    at Object.t.callErrorFromStatus (/azure-functions-host/workers/node/worker-bundle.js:2:40625)
    at Object.onReceiveStatus (/azure-functions-host/workers/node/worker-bundle.js:2:66215)
    at Object.onReceiveStatus (/azure-functions-host/workers/node/worker-bundle.js:2:58583)
    at /azure-functions-host/workers/node/worker-bundle.js:2:31694
    at processTicksAndRejections (node:internal/process/task_queues:78:11) {
  code: 14,
  details: 'Connection dropped',
  metadata: u { internalRepr: Map(0) {}, options: {} }
}
```
and then a
```
Error: 14 UNAVAILABLE: Connection dropped	
Error:
   at c.<anonymous> (at c.<anonymous> (/azure-functions-host/workers/node/worker-bundle.js:2:14000): /azure-functions-host/workers/node/worker-bundle.js:2)
   at c.emit (at c.emit (node:events:390:28): node:events:390)
   at c.emit (at c.emit (node:domain:475:12): node:domain:475)
   at Object.onReceiveStatus (at Object.onReceiveStatus (/azure-functions-host/workers/node/worker-bundle.js:2:66200): /azure-functions-host/workers/node/worker-bundle.js:2)
   at Object.onReceiveStatus (at Object.onReceiveStatus (/azure-functions-host/workers/node/worker-bundle.js:2:58583): /azure-functions-host/workers/node/worker-bundle.js:2)
   at <no_method> (at /azure-functions-host/workers/node/worker-bundle.js:2:31694: /azure-functions-host/workers/node/worker-bundle.js:2)
   at processTicksAndRejections (at processTicksAndRejections (node:internal/process/task_queues:78:11): node:internal/process/task_queues:78)
```

We can't see more output than this, logs before are just the expected logs from our code and after are all messages related to a new host startup.

Our setup is a durable function with fan-out-fan-in pattern: we take a pdf and then process each page (with poppler which has a lot of system dependencies) and finally wrap up the results.
Somewhere in one of the page processes a very hard error or some host issue or maybe an unreported timeout occurs, killing the nodejs instance / host and thus also killing the other fan-out processes running there. We see those processes re-appear (with IsReplay: false, which is kind of odd I think?) later on the new host which either gets killed as well or just goes through without issues. Eventually it will finish the whole process. This inconsistency makes it very hard to debug, mostly because we are flying blind: we just don't know what causes these errors... we are now in the processes of wrapping up as much as possible in try/catches, as we do assume we are doing something wrong. It's our first Azure function and the durable functions seem to be quite exotic - as in: not much examples to be found). It is a port of an aws lambda function which did the same but without the durable functions of course. We depend on Poppler, which we add with all libraries that are needed by means of the LD_LIBRARY_PATH env var. We extracted those libs from a local docker instance of the node runtime. This is quite tricky, but the only way to do this on linux consumption plan. If needed we can move to dedicated plan, but that kind of defeats the purpose as we have very unregular workloads so consumption plan is the best fit.

We do see this, after such a restart of the host while it is initializing:
```
Initializing Host. OperationId: '11d06b20-0899-470b-8dea-6dc4c5f6d285'.
Host initialization: ConsecutiveErrors=0, StartupCount=1, OperationId=11d06b20-0899-470b-8dea-6dc4c5f6d285
from grpc._cython import cygrpc as _cygrpc
File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
return future.result()
Traceback (most recent call last):
File "/usr/local/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
Traceback (most recent call last):
ImportError: cannot import name 'cygrpc' from 'grpc._cython' (/azure-functions-host/workers/python/3.6/LINUX/X64/grpc/_cython/__init__.py)
return aio_compat.run(start_async(
File "/azure-functions-host/workers/python/3.6/LINUX/X64/azure_functions_worker/dispatcher.py", line 19, in <module>
File "/azure-functions-host/workers/python/3.6/LINUX/X64/azure_functions_worker/main.py", line 50, in main
return loop.run_until_complete(main)
File "/azure-functions-host/workers/python/3.6/LINUX/X64/grpc/__init__.py", line 23, in <module>
File "/azure-functions-host/workers/python/3.6/LINUX/X64/azure_functions_worker/main.py", line 58, in start_async
from . import dispatcher
import grpc
return _run_code(code, main_globals, None,
return loop.run_until_complete(main)
File "/azure-functions-host/workers/python/3.6/LINUX/X64/azure_functions_worker/main.py", line 58, in start_async
File "/azure-functions-host/workers/python/3.6/LINUX/X64/azure_functions_worker/main.py", line 50, in main
ImportError: cannot import name 'cygrpc' from 'grpc._cython' (/azure-functions-host/workers/python/3.6/LINUX/X64/grpc/_cython/__init__.py)
from grpc._cython import cygrpc as _cygrpc
File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
return aio_compat.run(start_async(
File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
from . import dispatcher
exec(code, run_globals)
File "/azure-functions-host/workers/python/3.6/LINUX/X64/azure_functions_worker/__main__.py", line 6, in <module>
File "/azure-functions-host/workers/python/3.6/LINUX/X64/azure_functions_worker/dispatcher.py", line 19, in <module>
main.main()
return future.result()
File "/azure-functions-host/workers/python/3.6/LINUX/X64/grpc/__init__.py", line 23, in <module>
import grpc
File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
File "/usr/local/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
```
each one of those lines is a record on its own in application insights logs (this is dump of the message column from a csv export from those logs in the order it is received)
after this the host is started and works, so I'm not sure if it has anything to do with the errors before it, and if it does we still can't do anything about it

So in short our issue is that node workers just die without showing a cause.

**Comment by ejizba** [comment]
I'm cleaning up issues - sorry we never got to this, but hopefully you've either figured out your issue or moved on. Usually these "-1073740791" (aka "0xC0000409") errors are related to CPU/memory problems. We have a couple issues to track improvements in this area here (with potential workarounds listed):
- https://github.com/Azure/azure-functions-nodejs-library/issues/158
- https://github.com/Azure/azure-functions-nodejs-library/issues/161

If you're still having issues, please create a new issue or follow the steps in our new [troubleshooting guide](https://learn.microsoft.com/en-us/azure/azure-functions/functions-node-troubleshoot?pivots=nodejs-model-v4). Sorry again!
